kaos/algorithm/interpolator.py

```diff
@@ -0,0 +1,114 @@
+                                        .first())
+
+            # pylint: enable=undefined-variable
+        except NoResultFound:
+            return None
+
+    @staticmethod
+    def get_orbit_records_by_segment(segment_id):
+        """Get all the orbit records in a segment.
+
+        Args:
+            segment_id: The segment id.
+
+        Return:
+            All the orbit records, sorted by time.
+        """
+        # TODO move this to model
+        try:
+            # pylint: disable=undefined-variable
+            orbit_records = (OrbitRecords.query.filter_by(segment_id=segment_id)
+                                               .order_by(OrbitRecords.time))
+            # pylint: enable=undefined-variable
+            return list(orbit_records)
+        except NoResultFound:
+            return []
+        
+    @staticmethod
+    def linear_interp(platform_id, timestamp):
+        """Use linear interpolation to estimate the position and velocity of a satellite 
+        at a given time.
+
+        Args:
+            platform_id: The UID of the satellite.
+            timestamp: The time for which to get the estimated position and velocity.
+
+        Return:
+            A tuple (pos, vel). Both pos, vel are 3-tuples containing the vector components
+            of the position and velocity. Return None if interpolation could not be done
+            for the given satellite and timestamp.
+        """
+        # find the correct segment
+        segment = Interpolator.get_segment(platform_id, timestamp)
+        if not segment:
+            return None
+
+        orbit_records = Interpolator.get_orbit_records_by_segment(segment.segment_id)
+
+        # can't do linear interpolation in this case
+        if not orbit_records or len(orbit_records) < 2:
+            return None
+        
+        xp = np.array([rec.time for rec in orbit_records])
+        pos = np.array([np.array(rec.position) for rec in orbit_records])
+        vel = np.array([np.array(rec.velocity) for rec in orbit_records])
+
+        # TODO there's probably an elegant 1-liner way to do this with numpy
+        tpos = np.zeros(3)
+        tvel = np.zeros(3)
+
+        for i in range(3):
+            tpos[i] = np.interp(x=timestamp, xp=xp, fp=pos[:, i])
+            tvel[i] = np.interp(x=timestamp, xp=xp, fp=vel[:, i])
+
+        return tuple(tpos), tuple(tvel)
+
+
+    def interpolate(self, target):
+        """Estimate the position and velocity of the satellite at a given time.
+
+        Args:
+            target: The time for which to get the estimated position and velocity.
+
+        Return:
+            A tuple (pos, vel). Each of pos, vel is a 3-tuple representing the vector
+            components of the position and velocity, respectively. If interpolation could
+            not be done, return None.
+        """
+        return Interpolator.linear_interp(self.platform_id, target)
```

Comment:
Out of curiosity,  why does this method exist if it just calls another static method? Maybe you plan on expanding that in the future but as it stands I dont see a reason for the class.

Reply:
In the future, it would be useful to have a class that caches segment data so that multiple interpolations can be done at the same time.

I separated interpolate from linear_interp because it makes it easier to swap when we start experimenting with different interpolations.

Reply:
That seems like overkill since the interpolation is fine to just be two separate static methods. But I'll consider doing that if the lagrange becomes unwieldy.

Reply:
In that case LGTM